server/src/server.js

```diff
@@ -1111,33 +1118,52 @@ app.get("/api/fxa-oauth/confirm-login", function(req, res, next) {
   const data = state.split(END_POINT_SEPARATOR, 2);
   const endpoint = data[1];
 
-  checkState(req.deviceId, data[0]).then(isValid => {
+  const cookies = new Cookies(req, res, {keys: dbschema.getKeygrip("auth")});
+  const stateId = cookies.get("fxaState") || req.deviceId;
+  console.log("trying to verify with state", stateId, data[0]);
+  checkState(stateId, data[0]).then(isValid => {
+    console.log("after checkState", isValid);
     if (!isValid) {
       throw errors.badState();
     }
     return tradeCode(code);
   }).then(({ access_token: accessToken }) => {
+    console.log("after tradeCode", accessToken);
     return getAccountId(accessToken).then(({ uid: accountId }) => {
-      return registerAccount(req.deviceId, accountId, accessToken).then(() => {
-        return fetchProfileData(accessToken).then(({ avatar, displayName, email }) => {
-          return saveProfileData(accountId, avatar, displayName, email);
-        }).then(() => {
-          if (config.gaId) {
-            const analytics = ua(config.gaId);
-            analytics.event({
-              ec: "server",
-              ea: "fxa-login",
-              ua: req.headers["user-agent"],
-            }).send();
-          }
-          // Redirect to endpoint with auth param indicating successful Fxa auth flow.
-          // 'auth' param is used in reactrender and reactruntime to load wantsauth.js
-          // and display Fxa SignIn button state when request doesn't have accountId
-          // right after fxa-ouath/confirm-login redirection.
-          const pageUri = endpoint ? "/" + endpoint : "/";
-          res.redirect(pageUri + "?auth=1");
-        });
-      }).catch(next);
+      console.log("after accountId", accountId);
+      let promise = Promise.resolve();
+      if (req.deviceId) {
+        promise = registerAccount(req.deviceId, accountId, accessToken).then(() => {
+          return fetchProfileData(accessToken).then(({ avatar, displayName, email }) => {
+            return saveProfileData(accountId, avatar, displayName, email);
+          }).then(() => {
+            if (config.gaId) {
+              const analytics = ua(config.gaId);
+              analytics.event({
+                ec: "server",
+                ea: "fxa-login",
+                ua: req.headers["user-agent"],
+              }).send();
+            }
+            // Redirect to endpoint with auth param indicating successful Fxa auth flow.
+            // 'auth' param is used in reactrender and reactruntime to load wantsauth.js
+            // and display Fxa SignIn button state when request doesn't have accountId
+            // right after fxa-ouath/confirm-login redirection.
+            const pageUri = endpoint ? "/" + endpoint : "/";
+            // FIXME:
+            // We should be setting the accountid cookie in both paths, like this:
+            const cookies = new Cookies(req, res, {keys: dbschema.getKeygrip("auth")});
```

Comment:
Arguably we should use sendAuthInfo in both cases

Reply:
sendAuthInfo checks for deviceId before settings cookies, if we are going to use that flow we need to update that check to include accountId